KIPs/kip-149.md

```diff
@@ -0,0 +1,340 @@
+
+This standard defines the separation of data and logic contracts in system contracts. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized issue and potential private key loss.
+
+### Definitions
+
+- System contract: A contract that affects protocol. The currently deployed system contracts are as follows: **AddressBook**, **Governance**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **StakingTracker**, **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))
+
+- System contract upgrade: Upgrade a logic contract while maintaining a proxy contract. The upgraded logic contract must maintain the previous interface.
+
+- System contract replacement: Deploy a new system contract and register it to the Registry with the same name. It will deprecate predecessor system contract.
+
+### System Contracts Life Cycle
+
+The Registry contract manages the state of the system contract based on the current block number. The state transition cannot be reversed to the previous state; it only changes in one direction, as shown in the figure below.
+
+![](../assets/kip-149/LifeCycle.png)
+
+1. System Contract Deployed: It has been deployed. It contains connecting proxy and logic contracts and transferring ownership of proxy contract to Governance.
+
+2. Data Setup: It has all the data to be activated.
+
+3. Governance: Its registration has been proposed to Governance and is waiting to be registered if it passes.
+
+4. Registered: It has been registered in the registry but has not been activated yet.
+
+5. Active: It has been activated in the registry and is currently used.
+
+6. Deprecated: It has been deprecated in the registry and is no longer being managed or used.
+
+#### Upgrade System Contracts
+
+When upgrading system contracts, its logic contract will be changed by governance proposal. The Registry will not be updated since it only manages the address of the proxy contract.
+
+![](../assets/kip-149/UpgradeProcess.png)
+
+#### Replace System Contracts
+
+If system contract updates can’t be done by upgrading the logic contract, it must be replaced with the newly deployed contract. It needs two transactions from the governance: 1. Deprecate the predecessor system contract first and 2. Register the new system contract.
+
+The `successor's activation and the predecessor's deprecation` might differ, but the successor's `activation` block must be higher than the `deprecation` block of the predecessor.
```

Comment:
It was a legacy part, so you can dismiss this. Now the Registry doesn't hold `deprecation`. A predecessor system contract can only be deprecated by a new active successor. (It's almost the same as `GovParam`)

Reply:
I pointed out the operation misuse. For instance,

- [legacy contract]  activate (500) -> deprecate (zero address)
- [new contract]      registered -> activate (600)


At block number 550, the activation address is the legacy contract address. However, if the register function is called with the legacy contract name, zero address, and activation 550, the range from 550 to 559 returns a zero address, indicating no active contract for this range.

This operation isn't a bug, but it could lead to unintended consequences due to API misuse. Although this scenario is expected to be rare, I've asked the need for clarification or enhancement to address this potential issue.

Reply:
Could you give me some examples of `unintended consequences`? I'm trying to understand that what kinds of clarification or enhancement are needed other than only returning the result itself (return zero address)